modules/EvseManager/IECStateMachine.cpp

```diff
@@ -0,0 +1,370 @@
+        signal_event(std::get<CPEvent>(event));
+    }
+}
+
+void IECStateMachine::feed_state_machine() {
+    auto events = state_machine();
+
+    // Process all events
+    while (not events.empty()) {
+        signal_event(events.front());
+        events.pop();
+    }
+}
+
+// Main IEC state machine. Needs to be called whenever:
+// - CP state changes (both events from hardware as well as duty cycle changes)
+// - Allow power on changes
+// - The C1 6s timer expires
+std::queue<CPEvent> IECStateMachine::state_machine() {
+
+    std::queue<CPEvent> events;
+    std::lock_guard lock(state_mutex);
+
+    switch (cp_state) {
+
+    case RawCPState::Disabled:
+        if (last_cp_state != RawCPState::Disabled) {
+            pwm_running = false;
+            r_bsp->call_pwm_off();
+            ev_simplified_mode = false;
+            timeout_state_c1.stop();
+            call_allow_power_on_bsp(false);
+        }
+        break;
+
+    case RawCPState::A:
+        if (last_cp_state != RawCPState::A) {
+            pwm_running = false;
+            r_bsp->call_pwm_off();
+            ev_simplified_mode = false;
+            car_plugged_in = false;
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+        }
+
+        // Table A.6: Sequence 2.1 Unplug at state Bx (or any other
+        // state) Table A.6: Sequence 2.2 Unplug at state Cx, Dx
+        if (last_cp_state != RawCPState::A && last_cp_state != RawCPState::Disabled) {
+            events.push(CPEvent::CarRequestedStopPower);
+            events.push(CPEvent::CarUnplugged);
+        }
+        break;
+
+    case RawCPState::B:
+        // Table A.6: Sequence 7 EV stops charging
+        // Table A.6: Sequence 8.2 EV supply equipment
+        // responds to EV opens S2 (w/o PWM)
+
+        if (last_cp_state != RawCPState::A && last_cp_state != RawCPState::B) {
+
+            events.push(CPEvent::CarRequestedStopPower);
+            // Need to switch off according to Table A.6 Sequence 8.1
+            // within 100ms
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+        }
+
+        // Table A.6: Sequence 1.1 Plug-in
+        if (last_cp_state == RawCPState::A || last_cp_state == RawCPState::Disabled ||
+            (!car_plugged_in && last_cp_state == RawCPState::F)) {
+            events.push(CPEvent::CarPluggedIn);
+            ev_simplified_mode = false;
+        }
+
+        if (last_cp_state == RawCPState::E || last_cp_state == RawCPState::F) {
+            // Triggers SLAC start
+            events.push(CPEvent::EFtoBCD);
+        }
+        break;
+
+    case RawCPState::D:
+        // If state D is not supported switch off.
+        if (not has_ventilation) {
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+            break;
+        }
+        // no break, intended fall through: If we support state D it is handled the same way as state C
+    case RawCPState::C:
```

Comment:
During testing with a Metrel tester, we made a fast switch between A -> B -> C before the authorization is done. We got the following: 
Ignoring CarRequestedPower in WaitingForAuth state. Probably a BCB toggle that does not make sense here.

What we were expecting is: After authoziation has succeeded and the duty cycle was set to nominal, we would still be able to start charging, but we never got the allow_power_on(true).

Reply:
This has been fixed in main recently